File: c/simple2.c
#include <math.h>
#include <stdio.h>
#include <stdlib.h>

#define M 160
#define N 100
#define P 60
#define C 3
#define F 2
#define Q 100

const int Sx = M / 2;
const int Sy = N / 2;
const int Sz = P / 2;

void ramped_sin(float omega, float width, float delay, size_t q,
                float *result) {
  float t;
  for (size_t i = 0; i < q; ++i) {
    t = i * omega;
    result[i] = ((1.0 + tanh(t / width - delay)) / 2.0) * sin(t);
  }
}

size_t ix(int m, int n, int p, int c, int f) {
  return ((((m + M) % M * N + (n + N) % N) * P + (p + P) % P) * C + c) * F + f;
}

int main() {
  const float Sc = 0.99 / sqrt(3.0);

  float *EH = (float *)calloc(M * N * P * C * F, sizeof(float));

  float St[Q];
  ramped_sin(0.3, 5.0, 3.0, Q, St);

  size_t f, g;
  int s, m, n, p, c0, c1, c2, j0, j1, j2;

  for (size_t i = 0; i < 2 * Q; ++i) {
    f = i % 2;
    g = 1 - f;
    s = 1 - (2 * f);

    for (m = 0; m < M; ++m) {
      for (n = 0; n < N; ++n) {
        for (p = 0; p < P; ++p) {
          for (c0 = 0; c0 < 3; ++c0) {
            c1 = (c0 + 1) % 3;
            c2 = (c0 + 2) % 3;
            j0 = s * (c0 == 0);
            j1 = s * (c0 == 1);
            j2 = s * (c0 == 2);
            EH[ix(m - f, n - f, p - f, c0, f)] +=
                Sc * (EH[ix(m - f, n - f, p - f, c2, g)] -
                      EH[ix(m - j2 - f, n - j0 - f, p - j1 - f, c2, g)] -
                      EH[ix(m - f, n - f, p - f, c1, g)] +
                      EH[ix(m - j1 - f, n - j2 - f, p - j0 - f, c1, g)]);
          }
        }
      }
    }
    if (f == 0) {
      EH[ix(Sx, Sy, Sz, 2, 0)] += St[i / 2];
    }
  }

  FILE *file = fopen("output.bin", "wb");
  fwrite(EH, sizeof(float), 2 * M * N * P * 3, file);
  fclose(file);

  free(EH);

  return 0;
}

File: examples/simple2.rs
use anyhow::Result;
use std::fs::File;
use std::io::Write;

const M: usize = 160;
const N: usize = 100;
const P: usize = 60;
const C: usize = 3;
const F: usize = 2;
const Q: usize = 100;

const Sx: usize = M / 2;
const Sy: usize = N / 2;
const Sz: usize = P / 2;

fn ramped_sin(omega: f32, width: f32, delay: f32, q: usize) -> Vec<f32> {
    let mut result = Vec::with_capacity(q);
    for i in 0..q {
        let t = i as f32 * omega;
        let value = ((1.0 + (t / width - delay).tanh()) / 2.0) * t.sin();
        result.push(value);
    }
    result
}

#[inline]
fn ix(m: usize, dm: isize, n: usize, dn: isize, p: usize, dp: isize, c: usize, f: usize) -> usize {
    let mm = ((m as isize + dm + M as isize) % (M as isize)) as usize;
    let nn = ((n as isize + dn + N as isize) % (N as isize)) as usize;
    let pp = ((p as isize + dp + P as isize) % (P as isize)) as usize;
    return (((f * M + mm) * N + nn) * P + pp) * C + c;
}

fn main() -> Result<()> {
    let Sc = 0.99 / (3.0_f32).sqrt();
    let mut EH = vec![0.0_f32; M * N * P * C * F];
    let St = ramped_sin(0.3, 5.0, 3.0, Q);

    let mut c1: usize;
    let mut c2: usize;
    let mut f: isize;
    let mut g: isize;
    let mut s: isize;
    let mut j0: isize;
    let mut j1: isize;
    let mut j2: isize;

    for i in 0..(2 * Q) {
        f = (i % 2) as isize;
        g = 1 - f;
        s = 1 - (2 * f as isize);

        for m in 0..M {
            for n in 0..N {
                for p in 0..P {
                    for c0 in 0..C {
                        c1 = (c0 + 1) % 3;
                        c2 = (c0 + 2) % 3;
                        j0 = s * ((c0 == 0) as isize);
                        j1 = s * ((c0 == 1) as isize);
                        j2 = s * ((c0 == 2) as isize);
                        EH[ix(m, -f, n, -f, p, -f, c0, f as usize)] += Sc
                            * (EH[ix(m, -f, n, -f, p, -f, c2, g as usize)]
                                - EH[ix(m, -j2 - f, n, -j0 - f, p, -j1 - f, c2, g as usize)]
                                - EH[ix(m, -f, n, -f, p, -f, c1, g as usize)]
                                + EH[ix(m, -j1 - f, n, -j2 - f, p, -j0 - f, c1, g as usize)]);
                    }
                }
            }
        }
        if f == 0 {
            EH[ix(Sx, 0, Sy, 0, Sz, 0, 2, 0)] += St[i / 2];
        }
    }

    // Write output to binary file
    let mut file = File::create("output.bin")?;

    // Convert f32 slice to bytes and write
    let byte_data: Vec<u8> = EH[..2 * M * N * P * 3]
        .iter()
        .flat_map(|&f| f.to_le_bytes())
        .collect();

    file.write_all(&byte_data)?;

    Ok(())
}
